### PR TITLE
Improve asset data

### PR DESCRIPTION
**Pull request by joaoburatto**
## Summary

**Documentation:** https://ss3d.gitbook.io/systems/asset-data

Removes the need of `AddressableReferences` from the `AssetData` system, aswell as renaming it to `Assets`. This PR also automatizes the whole asset setup process, as the developer now only has to select which asset group he wants to create that asset list from, and click on "Load assets from addressables group", and then follow the same process as before, clicking on "Create Enum".

### Testing:
- Check if all addressable items are in the correct addressables group in the addressables's group window.,
- Check the `Project Settings` window, and see if the `Load All Asset Databases` button is functioning correctly.
- Check if all the `Asset Databases` in the project have all the assets loaded properly when pressing the `Load Assets From Addressables Group` window.
- Check if `Assets` uses are working normally on the built game.

### CODE REVIEW

**Review by stilnat — CHANGES_REQUESTED** [review]
Please modify all files and folders name to respect CamelCase, if some can't because they're external sources or something, please explain why in your PR.

**Comment by stilnat** [comment]
> Check if all the Asset Databases in the project have all the assets loaded properly when pressing the Load Assets From Addressables Group window.

Can you write a test for it ? Generally speaking, can you write a few tests for the whole thing ? Genuinely asking, if that's not relevant for some reason please explain why.

**Comment by stilnat** [comment]
> - Check if all the Asset Databases in the project have all the assets loaded properly when pressing the Load Assets From Addressables Group window.
> - Check if Assets uses are working normally on the built game.

Can you detail how to do that ?

**Comment by cosmiccoincidence** [comment]
> Please modify all files and folders name to respect CamelCase, if some can't because they're external sources or something, please explain why in your PR.

PascalCase not camelCase, haha

**Comment by stilnat** [comment]
> > Please modify all files and folders name to respect CamelCase, if some can't because they're external sources or something, please explain why in your PR.
> 
> PascalCase not camelCase, haha

Same thing !

**Comment by joaoburatto** [comment]
@stilnat @cosmiccoincidence lowerCamelCase and lowerPascalCase are the same thing

**Comment by joaoburatto** [comment]
@stilnat it is better to keep TextMeshPro's folder untouched.

**Comment by joaoburatto** [comment]
@stilnat @cosmiccoincidence ready to review again I think.

**Comment by joaoburatto** [comment]
@stilnat please ask for more tests if you notice something lacking, these are my first unit tests.

File: Assets/Scripts/SS3D/Systems/Storage/Items/Item.cs [hunk 1]
```diff
@@ -28,10 +28,11 @@ namespace SS3D.Systems.Storage.Items
     [RequiredLayer("Items")]
     public class Item : InteractionSource, IInteractionTarget
     {
+        [FormerlySerializedAs("ItemID")]
         [Header("Item settings")]
         [SerializeField]
         [HideInInspector]
-        public ItemIDs ItemID;
+        public Data.Enums.ItemIds _itemIdID;
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Storage/Items/Item.cs]: rename ItemIdId

File: Assets/Scripts/SS3D/Systems/Storage/Items/ItemSystem.cs [hunk 2]
```diff
@@ -33,22 +29,18 @@ protected override void OnStart()
         /// <summary>
         /// Loads the item prefabs into memory with the item id and setting up the item ids variable in the items.
         /// </summary>
-        private async void LoadItemPrefabs()
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Storage/Items/ItemSystem.cs]: what was the async for and why it's gone ?

File: Assets/Scripts/SS3D/Systems/Gamemodes/Objectives/GetItemObjective.cs [hunk 3]
```diff
@@ -21,12 +15,12 @@ public class GetItemObjective : GamemodeObjective
         /// <summary>
         /// The item id required to complete the objective.
         /// </summary>
-        [SerializeField] private ItemIDs _targetItemId;
+        [FormerlySerializedAs("_targetItemId")] [SerializeField] private Data.Enums.ItemIds _targetItemIdId;
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Gamemodes/Objectives/GetItemObjective.cs]: what's the FormerlySerializedAs attribute ? What does this need it ?

File: Assets/Scripts/EditorTests/AssetDataTests.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        private AssetDatabaseSettings _assetDatabaseSettings;
+
+
+        [SetUp]
+        public void SetUp()
+        {
+            _assetDatabaseSettings = ScriptableSettings.GetOrFind<AssetDatabaseSettings>();
+        }
+
+        [TearDown]
```
Comment (stilnat) [inline Assets/Scripts/EditorTests/AssetDataTests.cs]: Do we need that ?

File: Assets/Scripts/EditorTests/AssetDataTests.cs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        [Test] 
+        public void IncludedAssetDatabasesDoNotContainNullObjects()
+        {
+            // ARRANGE
+            List<AssetDatabase> loadedAssetDatabases = _assetDatabaseSettings.IncludedAssetDatabases;
+
+            bool hasNullAssets = false;
+            Dictionary<AssetDatabase, List<int>> assetDatabasesNullRefIndexes = new();
+
+            void debugNullAssets()
```
Comment (stilnat) [inline Assets/Scripts/EditorTests/AssetDataTests.cs]: Put it out of the method and make it private for readability, add a comment

File: Assets/Scripts/EditorTests/AssetDataTests.cs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+                    hasNullAssets = true;
+                    assetDatabasesNullRefIndexes.Add(assetDatabase, new List<int>());
+
+                    assetDatabasesNullRefIndexes.TryGetValue(assetDatabase, out List<int> assetIndexes);
+                    assetIndexes!.Add(index);
+                }
+            }
+
+            // DEBUG
```
Comment (stilnat) [inline Assets/Scripts/EditorTests/AssetDataTests.cs]: don't need that, the code is explicit

File: Assets/Scripts/EditorTests/AssetDataTests.cs [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            if (hasNullAssets)
+            {
+                debugNullAssets();
+            }
+
+            // ASSERT
```
Comment (stilnat) [inline Assets/Scripts/EditorTests/AssetDataTests.cs]: Don't need that the code is explicit

File: Assets/Scripts/EditorTests/AssetDataTests.cs [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+                    }
+
+                    foreach (int assetIndex in assetIndexes)
+                    {
+                        Debug.Log($"Asset is null on {assetDatabase.name} at index {assetIndex}");
+                    }
+                }
+            }
+
+            // ACT
```
Comment (stilnat) [inline Assets/Scripts/EditorTests/AssetDataTests.cs]: make a more explicit comment

File: Assets/Scripts/SS3D/Data/AssetDatabases/AssetDatabase.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+using UnityEngine;
+
+#if UNITY_EDITOR
+using UnityEditor.AddressableAssets.Settings;
+#endif
+
+namespace SS3D.Data.AssetDatabases
+{
+    /// <summary>
+    /// Generic database class, used to create asset databases.
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Data/AssetDatabases/AssetDatabase.cs]: Explain what's an asset database in the summary, as well as why we need that.

File: Assets/Scripts/SS3D/Data/AssetDatabases/AssetDatabase.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+            {
+                Assets.Add(entry.MainAsset);
+            }
+        }
+#endif
+
+        /// <summary>
+        /// Gets an asset based on its ID (index).
+        ///
+        /// WARNING: Not sure how
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Data/AssetDatabases/AssetDatabase.cs:66]: what do you mean not sure how ?

File: Assets/Scripts/SS3D/Data/AssetDatabases/InspectorEditor/AssetDatabaseInspectorEditor.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+        public VisualTreeAsset _assetDatabaseVisualTree;
+
+        private ScrollView _assetsListView;
+        private Button _loadAssetsButton;
+        private ObjectField _assetGroupObjectField;
+        private Label _assetDatabaseLabel;
+        private TextField _enumNameTextField;
+
+        private static readonly Regex SlashRegex = new(@"[\\//]");
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Data/AssetDatabases/InspectorEditor/AssetDatabaseInspectorEditor.cs]: I don't see that used anywhere

File: Assets/Scripts/SS3D/Data/AssetDatabases/InspectorEditor/AssetDatabaseInspectorEditor.cs [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+        private void OnEnable()
+        {
+            _assetDatabase = (AssetDatabase)target;
+        }
+
+        public override VisualElement CreateInspectorGUI()
+        {
+            VisualElement root = new();
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Data/AssetDatabases/InspectorEditor/AssetDatabaseInspectorEditor.cs]: Make a comment to explain what is expected to be displayed in inspector with this

File: Assets/Scripts/SS3D/Data/Assets.cs [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    /// Assets is used to load assets using enums or ints, with the addition of an enum generator for easy "id" of items.
+    /// </summary>
+    public static class Assets
+    {
+        /// <summary>
+        /// All loaded databases.
+        /// </summary>
+        private static readonly Dictionary<string, AssetDatabase> Databases = new();
+
+        // IMPORTANT: All database getters have to be added manually. For now.
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Data/Assets.cs]: What's the point of this comment ? Do we expect something different in the future ? If yes please add info about this.

File: Assets/Scripts/SS3D/Data/Assets.cs [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+        public static Sprite Get(InteractionIcons icon)
+        {
+            return GetDatabase(nameof(InteractionIcons)).Get<Sprite>((int)icon);
+        }
+
+        public static GameObject Get(ItemIds itemIdId)
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Data/Assets.cs]: Isn't there a better name for itemIdId ? It's not clear what it is from the name

Comment (stilnat) [inline Assets/Scripts/SS3D/Data/Assets.cs]: Add a comment, what game object does this return ?

File: Assets/Scripts/SS3D/Data/Assets.cs [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+    public static class Assets
+    {
+        /// <summary>
+        /// All loaded databases.
+        /// </summary>
+        private static readonly Dictionary<string, AssetDatabase> Databases = new();
+
+        // IMPORTANT: All database getters have to be added manually. For now.
+
+        public static Sprite Get(InteractionIcons icon)
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Data/Assets.cs:28]: Add a comment, what sprite does this return ?

File: Assets/Scripts/SS3D/Data/Assets.cs [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+        /// <typeparam name="TAssetType">The asset type you want returned.</typeparam>
+        /// <returns>The loaded asset in the TAssetType type.</returns>
+        public static TAssetType GetById<TAssetType>(string databaseName, int id)
+            where TAssetType : Object
+        {
+            return GetDatabase(databaseName).Get<TAssetType>(id);
+        }
+
+        /// <summary>
+        /// Loads the all asset databases on the project, for cache reasons.
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Data/Assets.cs]: Please precise the "for cache reasons" part. Loads all the, not Loads the all.

File: Assets/Scripts/SS3D/Attributes/ReadOnlyAttribute.cs [hunk 17]
```diff
@@ -0,0 +1,28 @@
+﻿#if UNITY_EDITOR
+using UnityEditor;
+using UnityEngine;
+
+namespace SS3D.Attributes
+{
+    public class ReadOnlyAttribute : PropertyAttribute { }
+
+    [CustomPropertyDrawer(typeof(ReadOnlyAttribute))]
+    public class ReadOnlyDrawer : PropertyDrawer
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Attributes/ReadOnlyAttribute.cs:16]: Make a summary explaining the reason why we need this class.

**Review by stilnat — CHANGES_REQUESTED** [review]

**Comment by stilnat** [comment]
Before merging, I'd like this (picture below) to be documented. I'd like to know what it is, and how to use it, for each options.
Please add it on the gitbook.

image

**Comment by TaylorNAlbarnaz** [comment]
> Before merging, I'd like this (picture below) to be documented. I'd like to know what it is, and how to use it, for each options. Please add it on the gitbook.
> 
> image

Yeah I agree, biggest problem with the Asset Data RN is having no idea how it works or how to use it and having to constantly ping John to ask him

File: Assets/Scripts/SS3D/Systems/Storage/Items/ItemSystem.cs (hunk 2 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Systems/Storage/Items/ItemSystem.cs]: Was used for the Addressables stuff, now we don't use it.

File: Assets/Scripts/SS3D/Systems/Gamemodes/Objectives/GetItemObjective.cs (hunk 3 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Systems/Gamemodes/Objectives/GetItemObjective.cs]: Used by Unity to not loose a prefab defined value.

File: Assets/Scripts/SS3D/Data/Enums/A.cs [hunk 18]
```diff
@@ -0,0 +1,17 @@
+namespace SS3D.Data.Enums
+{
+    public enum A
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Data/Enums/A.cs]: Why is this called just "A" ?

**Review by stilnat — DISMISSED** [review]

File: Assets/Scripts/SS3D/Data/Enums/A.cs (hunk 18 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Data/Enums/A.cs]: lol wtf

**Review by cosmiccoincidence — DISMISSED** [review]
### Rename to Pascal Case:
- Several files in `Assets/Art/Font/SpriteAssets/`
- Files in `Assets/Art/UI/`
- `/Art/Font/TextMeshPro/` back to `/Art/Font/TextMesh Pro/` this added like 300 files changes. Let's rename this in its own PR instead of bloating this one.

### Move:
- Font .asset files to `Assets/Content/Font/`   or maybe /Data/Font/
- `Assets/Art/UI/Editor/SS3D Editor.uss`  to the Content or Data folder.
- UI .asset files in `Assets/Art/UI/` to Content folder.

### Update:
- Merge in develop into the branch again and solve merge conflicts.

File: Assets/Scripts/SS3D/Systems/Storage/Items/Generic/BikeHorn.cs [hunk 19]
```diff
@@ -3,10 +3,9 @@
 using FishNet.Object;
 using SS3D.Interactions;
 using SS3D.Interactions.Interfaces;
-using SS3D.Systems.Storage.Items;
 using UnityEngine;
 
-namespace SS3D.Systems.Items.Generic
+namespace SS3D.Systems.Storage.Items.Generic
```
Comment (cosmiccoincidence) [inline Assets/Scripts/SS3D/Systems/Storage/Items/Generic/BikeHorn.cs]: Bikehorn goes under storage?

**Review by stilnat — COMMENTED** [review]
apart from the few comments here and there, I need to play test it and we're all good

**Review by stilnat — APPROVED** [review]
Good on my side ! I didn't thoroughly check folder change though but I trust you on this one.

**Comment by cosmiccoincidence** [comment]
I'm too lazy to double check this right now. As long as the new system is working as intended then I'm fine with it. Any file changes we missed can get picked up in a later PR.

**Pull request merged**